fix: use right gpu device for training when use device cuda:<index>

chudegao opened the pull request:
Not support train using gpu index as device. Use should set visible gpu device to use specific gpu before run ilab train command.


**Issue resolved by this Pull Request:**
Resolves #1279


**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

leseb commented on src/instructlab/lab.py:
```diff
@@ -876,6 +876,10 @@ def convert(self, value, param, ctx) -> "torch.device":
         # Third Party
+        if value.startswith("cuda:"):
+            index = value.split(":")[1]
+            os.environ["CUDA_VISIBLE_DEVICES"]=str(index)
```
Why is this env var needed?

chudegao replied:
in linux_train(): model = model.to(device) does not work to use specific gpu.

set environment CUDA_VISIBLE_DEVICES=<index> can make pytorch only see the <index> gpu.

refer https://github.com/huggingface/transformers/issues/25321

chudegao replied:
https://huggingface.co/docs/transformers/main/en/perf_train_gpu_many#order-of-gpus

tiran reviewed (changes requested):
Please set the env var in the `if device.type == "cuda"` block instead of parsing the input value manually.

The code should also handle the case of a pre-existing env var.

tiran commented:
I have reconsidered and now think that it would be better to not support device index at all. The approach has two flaws. 

- It is going to be hard to get it right when the user does something like `CUDA_VISIBLE_DEVICES=1,2` with `cuda:0`.
- AMD ROCm uses `--device=cuda`, too. I'm pretty sure you need a different env var on ROCm.

Let's refuse indexed CUDA devices and rather document how a user has to configure their system to select a specific hardware accelerator device.

chudegao commented:
> I have reconsidered and now think that it would be better to not support device index at all. The approach has two flaws.
> 
>     * It is going to be hard to get it right when the user does something like `CUDA_VISIBLE_DEVICES=1,2` with `cuda:0`.
> 
>     * AMD ROCm uses `--device=cuda`, too. I'm pretty sure you need a different env var on ROCm.
> 
> 
> Let's refuse indexed CUDA devices and rather document how a user has to configure their system to select a specific hardware accelerator device.

Agree. Let me update the code and doc.

tiran commented:
This would be a great opportunity to drastically simplify the code for `--device` parameter:

- drop `TorchDevice` class and replace it with a simple string option and `click.Choice(["cpu", "cuda", "hpu"])`
- pass device name as string to `linux_train()`
- perform CUDA check and HPU warning in `linux_train`

leseb commented on src/instructlab/lab.py:
```diff
@@ -972,12 +912,12 @@ def convert(self, value, param, ctx) -> "torch.device":
-        "PyTorch device for Linux training (default: 'cpu'). Use 'cuda' "
-        "for NVidia CUDA / AMD ROCm GPU, 'cuda:0' for first GPU."
+        "PyTorch device for Linux training. Use 'cuda' "
+        "for NVidia CUDA / AMD ROCm GPU, to use specific GPU, set visible GPU before run train command."
```
"set visible GPU before", how should users do that?

chudegao replied:
I think it's a little out of instructlab's scope. It's up to gpu type: NVidia CUDA / AMD ROCm GPU.  I suggest leave this to gpu user. The gpu user should know this well.
https://docs.nvidia.com/deploy/mps/#mps-on-multi-gpu-systems
https://rocm.docs.amd.com/en/latest/conceptual/gpu-isolation.html

chudegao replied:
I think it's a little out of instructlab's scope. It's up to gpu type: NVidia CUDA / AMD ROCm GPU.  I suggest leave this to gpu user. The gpu user should know this well.
https://docs.nvidia.com/deploy/mps/#mps-on-multi-gpu-systems
https://rocm.docs.amd.com/en/latest/conceptual/gpu-isolation.html

leseb commented on src/instructlab/train/linux_train.py:
```diff
@@ -149,10 +149,35 @@ def linux_train(
 ):
     """Lab Train for Linux!"""
+
+    device = torch.device(train_device)
```
Do you mind adding a note that we don't need error handling here given that we use `click.Choice(["cpu", "cuda", "hpu"])` earlier so if users don't pass the right value the CLI will fail? Thanks!

chudegao replied:
Let's add an error handling here in case linux_train() is called by other functions, the value maybe not use click.Choice(["cpu", "cuda", "hpu"]).

leseb commented on src/instructlab/train/linux_train.py:
```diff
@@ -149,10 +149,35 @@ def linux_train(
+        )
+
+    if four_bit_quant and device.type != "cuda":
+        ctx.fail("--4-bit-quant option requires --device=cuda")
```
```suggestion
        ctx.fail("'--4-bit-quant' option requires '--device=cuda'")
```

leseb reviewed (changes requested):
Given that this is somehow a regression, meaning we removed a functionality. I think we need to update CHANGELOG.md to surface this. Thanks!

leseb reviewed (changes requested):
Adding this again, given that this is somehow a regression, meaning we removed a functionality. I think we need to update CHANGELOG.md to surface this. Thanks!

chudegao commented:
> Adding this again, given that this is somehow a regression, meaning we removed a functionality. I think we need to update CHANGELOG.md to surface this. Thanks!

updated

leseb commented on CHANGELOG.md:
```diff
@@ -3,3 +3,5 @@
 
 ### Breaking Changes
+
+- **ilab model train** '--device' parameter does not support use gpu index(such as 'cuda:0') anymore.To use specific GPU,set visible GPU before run train command.
```
```suggestion
- `ilab model train`: The '--device' parameter no longer supports specifying a GPU index (e.g., 'cuda:0'). To use a specific GPU, set the visible GPU before running the train command.
```

chudegao replied:
updated

leseb reviewed (changes requested):
Please update your commit message too. Almost there :) Thanks!

leseb commented:
@chudegao see the CI failure:

```
py311-unitcov: commands[2]> .tox/py311-unitcov/bin/python -W error::UserWarning -m pytest --cov=instructlab --cov-report term --cov-report=html:coverage-py311-unitcov --cov-report=xml:coverage-py311-unitcov.xml --html=durations/py311-unitcov.html tests -m 'not (examples or slow)'
  ============================= test session starts ==============================
  platform linux -- Python 3.11.9, pytest-8.2.2, pluggy-1.5.0
  cachedir: .tox/py311-unitcov/.pytest_cache
  rootdir: /home/runner/work/instructlab/instructlab
  configfile: pyproject.toml
  plugins: asyncio-0.23.7, cov-5.0.0, anyio-4.4.0, html-4.1.1, metadata-3.1.1
  asyncio: mode=Mode.STRICT
  collected 56 items
  
  tests/test_config.py ........                                            [ 14%]
  tests/test_lab.py ...                                                    [ 19%]
  tests/test_lab_diff.py .............                                     [ 42%]
  tests/test_lab_download.py ..                                            [ 46%]
  tests/test_lab_generate.py ........                                      [ 60%]
  tests/test_lab_init.py ......                                            [ 71%]
  tests/test_lab_train.py .........F.                                      [ 91%]
  tests/test_notebooks.py .                                                [ 92%]
  /home/runner/work/instructlab/instructlab/.tox/py311-unitcov/lib/python3.11/site-packages/coverage/data.py:180: CoverageWarning: Data file '/home/runner/work/instructlab/instructlab/.coverage.fv-az1016-627.2781.XPTJRpgx' doesn't seem to be a coverage data file: 
    data._warn(str(exc))
  tests/test_utils.py ....                                                 [100%]
  
  =================================== FAILURES ===================================
  ________________________ TestLabTrain.test_train_linux _________________________
  
  self = <tests.test_lab_train.TestLabTrain object at 0x7f44fddf4b10>
  convert_llama_to_gguf_mock = <MagicMock name='convert_llama_to_gguf' id='1399[34](https://github.com/instructlab/instructlab/actions/runs/9482826056/job/26128541894?pr=1298#step:9:36)271393488'>
  linux_train_mock = <MagicMock name='linux_train' id='139934271403664'>
  is_macos_with_m_chip_mock = <MagicMock name='is_macos_with_m_chip' id='139934271288656'>
  
      @patch("instructlab.utils.is_macos_with_m_chip", return_value=False)
      @patch.object(linux_train, "linux_train")
      @patch(
          "instructlab.llamacpp.llamacpp_convert_to_gguf.convert_llama_to_gguf",
          side_effect=mock_convert_llama_to_gguf,
      )
      def test_train_linux(
          self,
          convert_llama_to_gguf_mock,
          linux_train_mock,
          is_macos_with_m_chip_mock,
      ):
          runner = CliRunner()
          with runner.isolated_filesystem():
              setup_input_dir()
              setup_linux_dir()
              result = runner.invoke(
                  lab.ilab,
                  ["--config=DEFAULT", "model", "train", "--input-dir", INPUT_DIR],
              )
              assert result.exit_code == 0
              convert_llama_to_gguf_mock.assert_called_once()
              assert convert_llama_to_gguf_mock.call_args[1]["model"] == Path(
                  "./training_results/final"
              )
              assert convert_llama_to_gguf_mock.call_args[1]["pad_vocab"] is True
              assert len(convert_llama_to_gguf_mock.call_args[1]) == 2
              linux_train_mock.assert_called_once()
              print(linux_train_mock.call_args[1])
              assert linux_train_mock.call_args[1]["train_file"] == Path(
                  "taxonomy_data/train_gen.jsonl"
              )
              assert linux_train_mock.call_args[1]["test_file"] == Path(
                  "taxonomy_data/test_gen.jsonl"
              )
              assert linux_train_mock.call_args[1]["num_epochs"] == 1
  >           assert linux_train_mock.call_args[1]["device"] is not None
  E           KeyError: 'device'
  
  tests/test_lab_train.py:[39](https://github.com/instructlab/instructlab/actions/runs/9482826056/job/26128541894?pr=1298#step:9:41)4: KeyError
  ----------------------------- Captured stdout call -----------------------------
  {'ctx': <click.core.Context object at 0x7f[44](https://github.com/instructlab/instructlab/actions/runs/9482826056/job/26128541894?pr=1298#step:9:46)fc880710>, 'train_file': PosixPath('taxonomy_data/train_gen.jsonl'), 'test_file': PosixPath('taxonomy_data/test_gen.jsonl'), 'model_name': 'instructlab/merlinite-7b-lab', 'num_epochs': 1, 'train_device': 'cpu', 'four_bit_quant': False}
```

chudegao commented:
I will check.

leseb commented on CHANGELOG.md:
```diff
@@ -3,3 +3,5 @@
 ### Features
 
 ### Breaking Changes
+
```
This needs to be changed, we now need to move this into the 0.18 section (please create it).

chudegao replied:
ok

leseb replied:
Actually I have a PR here https://github.com/instructlab/instructlab/pull/1427 so I stop asking people about it.

leseb replied:
The new release should be at the top of the file.

leseb commented:
@chudegao please rebase.
